File: src/cogsl/lex.go
/*markdown
# Lexical Analysis of CogSL (Cognitive Systems Language)

Consists of:

- Grammar: Declaring Syntax Rules
- Tokenization: Read in as stream of text, identify lexems & produce a stream of tokens
- Parsing: Read in the stream of tokens, assembling an AST

The design shall focus on streaming text, tokens & nodes.

The goal of CogSL is to express ideas & intent in a language whose form & function sits between natural language & programming languages.
CogSL should have a form closely related to natural language while providing functionality to rigidly structure language semantics.
To accomplish this, CogSL is self-defining:

- The lanuage defines a minimal set of grammar & syntax that it can use to describe itself.
- Users or Maintainers can extend or create supersets of the language by defining custom grammar & syntax.
- The defined grammar & syntax is extracted from the source for further semantic analysis.

## Grammar & Syntax

The reference implementation of CogSL uses a subset of English (herein lang) with emphasis on simple & straightforward prose that
follows well defined patterns. All grammar in lang must be defined for semantic meaning to be applied. Any grammar in the source is effectively ignored during analysis.

To be self describing, we first define a core set of language grammar & syntax:

- Sentence: A complete thought.
- Kinds of Sentences:
	- Declaratives: States a thought.
	- Imperatives: Authorative statement, commands or actions.
	- Conditionals: Expressing a particular set of circumstances
	- Interrogatives: Asks questions to clarify, cofirm or otherwise gather more information.
	- Obligations: Asserts a requirement or necessity.
	- Promises: statements of commitment to be eventually fullfilled
- Grammatical Structure of a Declarative Sentence:
	```
	SVO Sentence
	│
	├── Noun Phrase
	│		├── Noun
	├── Predicate Phrase
			├── Verb
			├── Object (Optional)
	```
  - Active Voice is the grammatical construction of choice.
		- Active Voice implements Subject-Verb-Object (SVO). The subject of the sentence performs the action expressed by the verb. The object recieves the action.
			- ex. She kicks the ball: `She` is the Subject, `kicks` is the verb, `the ball` is the object
	-
- Phrases, Clauses & Sentences



*/

package cogsl

import (
	"fmt"
	"os"
	"io"
	"unicode"
)

type TokenKind uint

const (
	TK_UNDEFINED TokenKind = iota // An undefined token
	TK_START_OF_DOC 						  // The start of the document
	TK_END_OF_DOC							    // The end of the document
	TK_WHITESPACE 							  // A collection of whitespace characters excluding newline unless it is escaped
	TK_LINE_BREAK 							  // An unescaped newline character
	TK_WORD 											// A collection of non-whitespace characters
)

type Token struct {
	Start int // The Starting index of the token in the buffer; ie. buffer[Start:Stop]
	Stop int // The Stopping index of the token in the buffer; ie. buffer[Start:Stop]
	Kind  TokenKind
}

type Tokenizer struct {
	TextSource io.Reader  // The Source of the text to analyze
	Buffer      []byte     // The buffer to write the source code into
	TokenSink  chan Token // The Channel on which to emit Lexical Tokens
}

func (t *Tokenizer) Tokenize(chunk_size int, growth_factor int, max_size int) error {
	/*
		Read the TextSource into the buffer. Assumes TextSource is a stream of UTF-8 characters from a single document.
		As characters are read, emit tokens accross the channel

		- chunk_size: The size of the chunk to read from the TextSource
		- growth_factor: How many chunk_sizes to grow the buffer when necessary

		# TODO

		- Use channels to control the tokenization process:
			- The Buffer becomes full
		- Dispatch the exact tokenization rules based on the state (FSM)
	*/
	growth_factor = max(growth_factor, 1)
	growth_size := chunk_size * growth_factor
	if t.Buffer == nil { return fmt.Errorf("the tokenizer's buffer was not initialized") }
	kind_log := make([]TokenKind, chunk_size)
	kind_log[0] = TK_START_OF_DOC
	t.TokenSink <- Token{0, 0, kind_log[0]}
	buf_iter_idx := 0 // where in the buffer are we currently
	escape_seq := false // Are we in an escape sequence?
	for {
		// Read a chunk of characters into the buffer
		chunk := make([]byte, chunk_size)
		n, err := t.TextSource.Read(chunk)
		fmt.Fprintf(os.Stderr, "Read %d bytes\n", n)
		if n == 0 && err == nil { continue } // TODO: Better handle this case; usually encountered in Non-Blocking IO
		if n > 0 && cap(t.Buffer) < len(t.Buffer) + n { // Grow the buffer if necessary
			extension_size := min(max_size - len(t.Buffer), growth_size)
			if extension_size == 0 { return fmt.Errorf("the tokenizer's buffer has reached the maximum size of %d", max_size) }
			t.Buffer = append(t.Buffer, make([]byte, 0, extension_size)...)
		}
		if err != nil {
			if err == io.EOF {
				fmt.Fprintf(os.Stderr, "Buffer Iter Idx: %d\nBuffer Length: %d\n", buf_iter_idx, len(t.Buffer))
				// If there is text left in the buffer emit it as a token of the last kind & then emit the end of document token
				if buf_iter_idx < len(t.Buffer) - 1 { t.TokenSink <- Token{buf_iter_idx, len(t.Buffer) - buf_iter_idx, kind_log[len(kind_log)-1]} }
				t.TokenSink <- Token{buf_iter_idx, buf_iter_idx, TK_END_OF_DOC}
				break
			} else {
				close(t.TokenSink); return err
			}
		}
		chunk_start_idx := len(t.Buffer) // Where the Chunk starts in the buffer; just the end of the buffer before we append the chunk
		t.Buffer = append(t.Buffer, chunk[:n]...)
		fmt.Fprintf(os.Stderr, "Chunk Start Idx: %d\n", len(t.Buffer))

		// Parse the chunk into tokens
		var rune_kind TokenKind
		chunk_iter_idx := 0 // where in the chunk are we currently
		for i := 0; i < n; i++ {
			cur_rune := rune(t.Buffer[i])
			switch {
			// Regular Cases
			case !escape_seq && cur_rune == '\\': escape_seq = true // A New Escape Sequence
			case !escape_seq && unicode.IsSpace(cur_rune): if cur_rune == '\n' { rune_kind = TK_LINE_BREAK } else { rune_kind = TK_WHITESPACE } // Whitespace
			case !escape_seq: rune_kind = TK_WORD // If it's not an escape sequence or whitespace, then it's a word
			// Escape Sequence Cases
			case escape_seq && cur_rune == '\n': escape_seq = false; rune_kind = TK_WORD // A newline escape sequence
			case escape_seq && cur_rune == '\\': escape_seq = false; rune_kind = TK_WORD // Escaping the escape character
			// There are no other valid escape sequences so emit an undefined token
			default: escape_seq = false; rune_kind = TK_UNDEFINED
			}
			if escape_seq { continue; } // If we are in an escape sequence then we short circuit & jump to the next rune
			// If the kind of rune is different from the previous, then evaluate if we need to emit a token
			last_kind := kind_log[len(kind_log) - 1]
			if rune_kind != last_kind {
				// Edge Triggered: Emit the token, append the new kind to the log & update our indices
				chunk_iter_idx = i
				token_start_idx := buf_iter_idx // The Token starts at the current buffer index
				token_stop_idx := chunk_start_idx + chunk_iter_idx // The Token stops at the position in the buffer relative to the start of the chunk
				t.TokenSink <- Token{token_start_idx, token_stop_idx, last_kind}
				kind_log = append(kind_log, rune_kind)
				buf_iter_idx = token_stop_idx // The New Buffer index is where the token stopped
			}
			// otherwise we just continue
		}
	}

	// Cleanup our end of the channel
	close(t.TokenSink)
	return nil
}

File: src/cogsl/lex.py
"""

Lexical Analyzer for the Cognitive Systems Language

"""
from __future__ import annotations
from typing import TextIO, Any, TypedDict, NotRequired, Literal, Generic, TypeVar
from collections.abc import Iterator, Iterable

import io
from . import LEXICON_SYMBOLS

token_kind_t = Literal['Word']

class DocumentToken(TypedDict):
  buffer: str
  """The Token itself"""
  pos: int
  """Where in the document the Token starts"""

class LexicalToken(TypedDict):
  """The Minimal set of attributes necessary for a lexical Token"""
  buffer: str
  """The Token itself"""
  kind: token_kind_t
  """The Classification of Token"""

  class Word(TypedDict):
    """The attributes expected of a Word"""
    kind: Literal['Word']
    """The Classification of Token"""
    reserved: bool
    """Denotes if this word is defined in the Lexicon"""

ID = TypeVar('ID')
NODE = TypeVar('NODE')
class Tree(Generic[ID, NODE], TypedDict):
  root_node: ID
  nodes: dict[ID, NODE]
  edges: dict[ID, set[ID]]

token_seq_t = Iterator[DocumentToken]
lexical_seq_t = Iterator[LexicalToken]
ast_t = Tree[int, Any]

CHUNK_SIZE = 16384

class Scanner:

  @staticmethod
  def _is_legal(symbol: str) -> bool:
    """Evaluates if the Symbol is legal"""
    raise NotImplementedError

  @staticmethod
  def scan(source: TextIO, chunk_size: int = CHUNK_SIZE) -> TextIO:
    """Reads a Text Source checking for illegal symbols"""
    buffer = io.StringIO()
    buf_pos = 0
    eof = False
    while not eof:
      chunk = source.read(chunk_size)
      eof = chunk < chunk_size
      for symbol in chunk:
        illegal = not Scanner._is_legal(symbol)
        if illegal: raise ValueError(f"Illegal Symbol `{symbol}` at Position `{buf_pos}` in the Stream")
        buffer.write(symbol)
        buf_pos += 1
    return buffer

class Tokenizer:

  class Document:

    @staticmethod
    def _split_on_symbols(buffer: str, symbols: set[str], max_split: int = None) -> list[tuple[str, str | None]]:
      """Splits the Buffer into items after each occurance of any one of the symbols.

      If `max_split` times is set then stop splitting after N matches

      returns the list of items as a tuple of (buffer, symbol)
      """
      raise NotImplementedError

    @staticmethod
    def tokenize(document: TextIO, chunk_size: int = CHUNK_SIZE) -> token_seq_t:
      """Convert a document into a Token Sequence IR"""
      stash = io.StringIO()
      tokens: list[DocumentToken] = []
      eof = False
      while not eof:
        chunk = document.read(chunk_size)
        eof = chunk < chunk_size
        buffer = stash.read() + chunk # Prepend the previous stash

        split_order = (
          (set(LEXICON_SYMBOLS['Whitespace']), None),
          (set(LEXICON_SYMBOLS['Punctuation']), None),
        )
        items = [ { 'buffer': buffer } ] # Start with a single item
        for _split_eval_idx, (symbols, max_split) in enumerate(split_order):
          """
          
          Here we take the buffer & split it on the symbols declared.
            The last most item is stashed for later evaluation if we haven't reached EoF since we don't know if the token extends into the next chunk
          We then further split each item on the next set of symbols repeating until all splitting symbols are evaluated.

          TODO: Add Document Positions & Other MetaData
          
          """
          _items = []
          for item in items:
            # For Each Item; split the item's buffer on the set of symbols
            if max_split is not None: subitems = Tokenizer._split_on_symbols(item['buffer'], symbols, max_split=max_split)
            else: subitems = Tokenizer._split_on_symbols(item['buffer'], symbols)
            
            # Stash the last Item if we haven't encountered EoF
            if not eof and _split_eval_idx == 0:
              _buf, _sep = subitems[-1]
              stash.write(_buf)
              if _sep is not None: stash.write(_sep)
              subitems = subitems[:-1]

            # Flatten the List
            for _buf, _sep in subitems:
              _items.append({ 'buffer': _buf })
              if _sep is not None: _items.append( { 'buffer': _sep } )

          # Swap out the current set of items to evaluate
          items = _items
      
      assert len(stash.getvalue()) == 0, "Stash has leftovers when it shouldn't"

      # TODO: Fill-in remaining Token Metadata

      return iter(tokens)

  class Lexicon:

    @staticmethod
    def _eval_ruleset(token_seq: Iterable[DocumentToken]) -> tuple[bool, LexicalToken | None]:
      """Attempts to evaluate the sequence of tokens into a single Lexical token.
      
      If evaluation was succesful then returns (True, LexicalToken)
      otherwise returns (False, None)
      """
      raise NotImplementedError

    @staticmethod
    def _lookup_word(word: LexicalToken) -> tuple[bool, LexicalToken.Word | None]:
      """Looks up the Word in the Lexicon's Dictionary"""
      assert word['kind'] == 'Word'
      raise NotImplementedError

    ### TODO: Probably need to revisit this once AST Assembly is implemented
    #
    #
    #
    @staticmethod
    def tokenize(token_seq: token_seq_t) -> lexical_seq_t:
      """Tokenize the DocumentToken Sequence IR against the Lexicon to generate a LexicalToken Sequence"""

      tokens: list[LexicalToken] = []
      tkn_stash: list[DocumentToken] = []
      for doc_tkn in token_seq:
        tkn_stash.append(doc_tkn)
        ok, lex_tkn = Tokenizer.Lexicon._eval_ruleset(tkn_stash)
        if not ok: continue
        tkn_stash.clear()
        tokens.append(lex_tkn)
        if lex_tkn['kind'] == 'Word':
          ok, word = Tokenizer.Lexicon._lookup_word(lex_tkn)
          if ok: tokens[-1] |= word
      
      return iter(tokens)
    #
    #
    #
    ###

class Parser:

  @staticmethod
  def _eval_format(seq: Iterable[LexicalToken]) -> tuple[bool, Any | None]:
    """Evaluates the Sequence of Lexical Tokens against the Formatting Rules & returns the largest possible formatting group"""
    raise NotImplementedError
  
  @staticmethod
  def _eval_prose(seq: Iterable[LexicalToken]) -> tuple[bool, Any | None]:
    """Evaluates the sequence of Lexical Tokens against the Prose Rules & returns the largest possible prose group"""
    raise NotImplementedError
  
  @staticmethod
  def parse(lex_seq: lexical_seq_t) -> ast_t:
    """Produce an Abstract Syntax Tree from the provided Lexical Sequence."""
    
    """NOTE: Proposed Implementation
    
    1. First, we evaluate the sequence against formatting rules grouping the tokens matching a specific rule.
      - Recursively Evaluate each grouping's sub-sequence against the formatting rules until no rules match.
    2. Next, for each formatting group, evaluate the lexical (sub)sequence against the prose rules.
      - Recursively Evaluate each grouping's sub-sequence against the prose rules un until no rules match.
      - At this point each Prose Group is considered "terminal syntax" & no further evaluation can occur.
    3. Assemble the AST walking the recursive evaluation hierarchy.

    """

    raise NotImplementedError